Services/UI/OJS.Services.Ui.Business/Cache/Implementations/ContestsCacheService.cs

```diff
@@ -1,48 +1,100 @@
+    private readonly IProblemsCacheService problemsCache;
 
     public ContestsCacheService(
         ICacheService cache,
-        IContestsDataService contestsData)
+        IContestsDataService contestsData,
+        IProblemsCacheService problemsCache)
     {
         this.cache = cache;
         this.contestsData = contestsData;
+        this.problemsCache = problemsCache;
     }
 
-    public async Task<ContestDetailsServiceModel?> GetContestDetailsServiceModel(
-        int contestId,
-        int cacheSeconds = CacheConstants.FiveMinutesInSeconds)
-        => await this.cache.Get(
-            string.Format(CacheConstants.ContestDetails, contestId),
-            async () => await this.GetContestServiceModel(contestId),
-            cacheSeconds);
-
-    private async Task<ContestDetailsServiceModel?> GetContestServiceModel(int contestId)
+    public async Task<ContestDetailsServiceModel?> GetContestDetailsServiceModel(int contestId)
     {
-        var contestDetailsServiceModel = await this.contestsData.GetById<ContestDetailsServiceModel>(contestId);
+        var (contest, problems) = await this.GetContestWithProblems(contestId);
 
-        if (contestDetailsServiceModel != null)
+        if (contest is null)
         {
-            contestDetailsServiceModel.AllowedSubmissionTypes = contestDetailsServiceModel.Problems
-                .SelectMany(p => p.AllowedSubmissionTypes)
-                .DistinctBy(st => st.Id)
-                .MapCollection<ContestDetailsSubmissionTypeServiceModel>()
-                .ToList();
+            return null;
         }
 
+        var contestDetailsServiceModel = contest.Map<ContestDetailsServiceModel>();
+
+        contestDetailsServiceModel.Problems = problems.MapCollection<ContestProblemServiceModel>().ToList();
+
+        contestDetailsServiceModel.AllowedSubmissionTypes = contestDetailsServiceModel.Problems
+            .SelectMany(p => p.AllowedSubmissionTypes)
+            .DistinctBy(st => st.Id)
+            .MapCollection<ContestDetailsSubmissionTypeServiceModel>()
+            .ToList();
+
         return contestDetailsServiceModel;
     }
+
+    public async Task<ContestServiceModel?> GetContestServiceModel(int contestId)
+    {
+        var (contest, problems) = await this.GetContestWithProblems(contestId);
+
+        if (contest is null)
+        {
+            return null;
+        }
+
+        var contestServiceModel = contest.Map<ContestServiceModel>();
+
+        contestServiceModel.Problems = problems.MapCollection<ContestProblemServiceModel>().ToList();
+
+        contestServiceModel.AllowedSubmissionTypes = contestServiceModel.Problems
+            .SelectMany(p => p.AllowedSubmissionTypes)
+            .DistinctBy(st => st.Id)
+            .ToList();
+
+        return contestServiceModel;
+    }
+
+    public async Task<(ContestCacheModel? Contest, ICollection<ProblemCacheModel> Problems)> GetContestWithProblems(int contestId)
```

Comment:
This method is redundant, as it just calls 2 other methods without any other modifications. It's more obvious to get them both where needed.